### PR TITLE
106 circuit builder for bbf

### PR DESCRIPTION
**Pull request by ayashunsky**
Merge into master the current working state of circuit_builder

### CODE REVIEW

File: crypto3/libs/blueprint/include/nil/blueprint/bbf/circuit_builder.hpp [hunk 1]
```diff
@@ -0,0 +1,668 @@
...
+                    const auto &dynamic_tables = bp.get_reserved_dynamic_tables();
+
+                    std::vector<std::string> ordered_table_names =
+                        nil::crypto3::zk::snark::get_tables_ordered_by_rows_number<FieldType>(lookup_tables);
+
+                    std::size_t start_row = 1;
+                    std::vector<plonk_lookup_table> bp_lookup_tables(lookup_table_ids.size());
+                    std::size_t start_constant_column = presets.constants_amount();
+                    std::size_t prev_columns_number = 0;
+                    int full_selector_id = 0;
```
Comment (martun) [inline crypto3/libs/blueprint/include/nil/blueprint/bbf/circuit_builder.hpp:235]: Are we sure 0 here is good enough a merker? Maybe change to -1? AFAIK 0 is a proper selector id.

File: crypto3/libs/blueprint/include/nil/blueprint/bbf/circuit_builder.hpp [hunk 2]
```diff
@@ -0,0 +1,668 @@
...
+                                row_selector selector_column(usable_rows);
+                                for(std::size_t i = 1; i < usable_rows; i++)
+                                    selector_column.set_row(i);
+                                full_selector_id = gates.add_selector(selector_column);
+                            }
+
+                            if (table->get_rows_number() % usable_rows == 0) options_number--;
+                            std::size_t cur = 0;
+                            for(std::size_t i = 0; i < options_number; i++) {
+                                for(std::size_t start_row = 1; start_row < usable_rows; start_row++, cur++) {
```
Comment (martun) [inline crypto3/libs/blueprint/include/nil/blueprint/bbf/circuit_builder.hpp:263]: You already have a variable named start_row, please don't reuse the name.

File: crypto3/libs/blueprint/include/nil/blueprint/bbf/circuit_builder.hpp [hunk 3]
```diff
@@ -0,0 +1,668 @@
...
+                                      << " second variable: " << copy_constraints[i].second << std::endl;
+                            std::cout << var_value(assignments, copy_constraints[i].first) << " != "
+                                      << var_value(assignments, copy_constraints[i].second) << std::endl;
+                            return false;
+                        }
+                    }
+                    return true;
+                }
+
+                bool check_proof(
```
Comment (martun) [inline crypto3/libs/blueprint/include/nil/blueprint/bbf/circuit_builder.hpp:601]: I believe this function belongs to the test, not this class. Blueprint probably must not be aware of the prover.

**Review by martun — COMMENTED** [review]
Added 3 comments.

File: crypto3/libs/blueprint/include/nil/blueprint/bbf/circuit_builder.hpp (hunk 1 above)
Comment (ayashunsky) [reply crypto3/libs/blueprint/include/nil/blueprint/bbf/circuit_builder.hpp]: Fixed. I had the same idea (wherefrom int is the type) but then reverted for some reason.

File: crypto3/libs/blueprint/include/nil/blueprint/bbf/circuit_builder.hpp (hunk 2 above)
Comment (ayashunsky) [reply crypto3/libs/blueprint/include/nil/blueprint/bbf/circuit_builder.hpp]: fixed

File: crypto3/libs/blueprint/include/nil/blueprint/bbf/circuit_builder.hpp (hunk 3 above)
Comment (ayashunsky) [reply crypto3/libs/blueprint/include/nil/blueprint/bbf/circuit_builder.hpp]: Moved function to test.

**Review by martun — APPROVED** [review]

**Pull request merged**